torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -691,370 +691,418 @@ def inner_fn(args: List[Any]):
+            # This is a bit subtle. The goal of this entire function (aot_dispatch_synthetic_bases)
+            # is to relieve the downstream logic from having to reason about mutations on inputs that alias
+            # each other, by replacing aliased inputs with a synthetic base.
+            # One area where this breaks down a bit however is if one of those aliased inputs
+            # experienced a metadata mutation.
+            # We are now obligated to reapply the metadata mutation directly to the user's input;
+            # it isn't enough to apply mutations back to the synthetic base in the downstream logic.
+            #
+            # The way we handle this is by pretending that those aliased inputs that experience metadata mutations
+            # are additional outputs in the user's forward function.
+            # The downstream logic will just treat these as "user outputs that alias inputs".
+            # However, we will manually grab them at runtime here, use them to reapply the metadata mutation
+            # to the user inputs, and not return them to the user.
+            aliased_args_with_metadata_mutations = [
+                x
+                for i, x in enumerate(unpacked_args)
+                if i in self.aliased_arg_idx_with_metadata_mutations
+            ]
+            if len(aliased_args_with_metadata_mutations) > 0:
+                return *(flat_fn(*unpacked_args)), *aliased_args_with_metadata_mutations
+            else:
+                return flat_fn(*unpacked_args)
+
+        if config.debug_assert:
+            ref_fw_metadata = run_functionalized_fw_and_collect_metadata(
+                wrapped_flat_fn,
+                keep_input_mutations=fw_metadata.keep_input_mutations,
+                is_train=fw_metadata.is_train,
+            )(*flat_args_with_synthetic_bases)
+            assert ref_fw_metadata == fw_metadata_updated, (
+                f"ref_metadata={pprint.pformat(partial_flatten_asdict(ref_fw_metadata))}, "
+                f"\nactual_metadata={pprint.pformat(partial_flatten_asdict(fw_metadata_updated))}"
+            )
+        return (
             wrapped_flat_fn,
-            keep_input_mutations=fw_metadata.keep_input_mutations,
-            is_train=fw_metadata.is_train,
-        )(*flat_args_with_synthetic_bases)
-        assert ref_fw_metadata == fw_metadata_updated, (
-            f"ref_metadata={pprint.pformat(partial_flatten_asdict(ref_fw_metadata))}, "
-            f"\nactual_metadata={pprint.pformat(partial_flatten_asdict(fw_metadata_updated))}"
+            flat_args_with_synthetic_bases,
+            aot_config,
+            fw_metadata_updated,
         )
 
-    compiled_fn = compiler_fn(
-        wrapped_flat_fn,
-        flat_args_with_synthetic_bases,
-        aot_config,
-        fw_metadata=fw_metadata_updated,
-    )
+    def post_compile(
+        self,
+        compiled_fn,
+        aot_config: AOTConfig,
+        *,
+        fw_metadata: ViewAndMutationMeta,
```

Comment:
Mostly just confirming, but it looks like:

(1) the fw_metadata that we pass in here is actually not very useful: it contains the version of the metadata **after** synthetic bases were handled, while we actually need the version of the metadata before they are handled to do the runtime logic

(2) we end up just not using it at all: in practice, the only metadata that this pass needs is the `fw_metadata.input_info`, so you just stash the "correct" one as extra metadata on the class, and we effectively ignore any info on `fw_metadata`.

Does that sound right to you? I guess this fits with every pass requiring a combo of "shared metadata" and "pass-specific metadata", and here it's the case that all the metadata we need is specific to this pass / stashed on the object.

Reply:
ah I see, @aorenste pointed out something similar

Reply:
Chatted with @aorenste  offline and came to a similar conclusion: 

Wrappers come in two categories:

1. Wrappers that modify fw_metadata, but then never use fw_metadata in post compile
2. Wrappers that never modify fw_metadata, and only consume the "final" fw_metadata. 

And category 1 always wraps around category 2, and they're split by a function boundary. 

So I think it's sufficient to change the invariant and say that you just aren't guaranteed that the fw_metadata passed to you in post compile is the one you modified: instead it's always the fw_metadata after applying all the wrappers. It happens that wrappers in category 1 never use fw_metadata at all today, but I can imagine a future where there's a wrapper that modifies fw_metadata, but still wants access to the "final" version of fw_metadata after all wrappers have run. 

In order to avoid merge conflicts, I put a renaming change in to the PR above this, where I rename fw_metadata in post_compile to "runtime_metadata", to clearly signify that it's a different metadata from the one passed in.